src/server/service/kv.rs

```diff
@@ -142,18 +144,27 @@ macro_rules! handle_request {
                 return;
             }
             let begin_instant = Instant::now_coarse();
-            let future = $future_name(&self.storage, req)
-                .and_then(|res| sink.success(res).map_err(Error::from))
-                .map(move |_| GRPC_MSG_HISTOGRAM_STATIC.$fn_name.observe(duration_to_sec(begin_instant.elapsed())))
-                .map_err(move |e| {
-                    debug!("kv rpc failed";
-                        "request" => stringify!($fn_name),
-                        "err" => ?e
-                    );
-                    GRPC_MSG_FAIL_COUNTER.$fn_name.inc();
-                });
 
-            ctx.spawn(future);
+            let resp = $future_name(&self.storage, req);
+            let task = async move {
+                let result = async move {
```

Comment:
Why create another closure?

Reply:
This approach unifies the results, and can also use `and_then()` to generate a chained future.

Reply:
Maybe I can use `future.map_ok().map_err()` to make this clearer

Reply:
I pushed new commit to use `future.map_ok().map_err()` to make this clearer